charms/k8s/src/charm.py

```diff
@@ -16,35 +16,143 @@
         self.framework.observe(self.on.update_status, self._on_update_status)
+        self.framework.observe(self.on.install, self._on_install)
 
-    def _on_update_status(self, _event: ops.UpdateStatusEvent):
+        factory = UnixSocketConnectionFactory(unix_socket=K8SD_SNAP_SOCKET)
+        self.api_manager = K8sdAPIManager(factory)
+        self.snap_cache = SnapCache()
+
+    def _apply_snap_requirements(self):
+        """Apply necessary snap requirements for the k8s snap.
+
+        This method executes necessary scripts to ensure that the snap
+        meets the network and interface requirements.
+        """
+        self.unit.status = ops.MaintenanceStatus("Applying K8s requirements")
+        commands = [
+            "/snap/k8s/current/k8s/connect-interfaces.sh",
+            "/snap/k8s/current/k8s/network-requirements.sh",
+        ]
+        for c in commands:
+            subprocess.check_call(shlex.split(c))
+
+    def _bootstrap_k8s_snap(self):
+        """Bootstrap the k8s if it's not already bootstrapped."""
+        if not self.api_manager.is_cluster_bootstrapped():
+            self.unit.status = ops.MaintenanceStatus("Bootstrapping Cluster")
+            cmd = "k8s bootstrap"
+            subprocess.check_call(shlex.split(cmd), shell=False)
+
+    def _enable_components(self):
+        """Enable necessary components for the Kubernetes cluster."""
+        self.unit.status = ops.MaintenanceStatus("Enabling DNS")
+        self.api_manager.enable_component("dns", True)
+        self.unit.status = ops.MaintenanceStatus("Enabling Network")
+        self.api_manager.enable_component("network", True)
+
+    def _get_snap_version(self) -> Optional[str]:
+        """Retrieve the version of the installed Kubernetes snap package.
+
+        Returns:
+            Optional[str]: The version of the installed k8s snap package, or None if
+            not available.
+        """
+        cmd = "snap list k8s"
+        result = subprocess.check_output(shlex.split(cmd))
+        output = result.decode().strip()
+        match = re.search(r"(\d+\.\d+(?:\.\d+)?)", output)
+
+        if match:
+            return match.group()
+
+        logger.info("Snap k8s not found or no version available.")
+        return None
+
+    def _install_k8s_snap(self):
+        """Install the k8s snap package."""
+        self.unit.status = ops.MaintenanceStatus("Installing k8s snap")
+        k8s_snap = self.snap_cache["k8s"]
+        if not k8s_snap.present:
+            k8s_snap.ensure(SnapState.Latest, channel="edge")
+
+    def _on_install(self, event):
+        """Handle install event for the charm.
+
+        Args:
+            event: The event that triggered this handler.
+        """
+        try:
+            self._install_k8s_snap()
+            self._apply_snap_requirements()
+            self._bootstrap_k8s_snap()
+            self._enable_components()
+            self.unit.status = WaitingStatus("K8s not ready")
```

Comment:
This line is a bit confusing. Why isn't this ops.WaitingStatus ? And why does it way that k8s is not ready? Are we waiting for the update hook to run to get the status updated to ready? If yes, then that would be after how long?

Reply:
I'll refactor this fragment. We should directly invoke a set_status handler and finish the event. If it's not ready, we will wait until the update_status handler is run. Regarding the periodicity of the update_status events, they should be fired at 5 min intervals; however, users can change this value.